### PR TITLE
Only perform config loading re-entrancy check for cjs

### LINKED ISSUE #15916: [Bug]: @babel/core version 7.22.11 breaks the build
### How are you using Babel?

@rollup/plugin-babel

### Input code

I have a simple repro in this repo https://github.com/arthirm/babel-core-issue1. Downloading and doing `yarn install` and `yarn build` should break the build with the following error

```
[!] (plugin babel) SyntaxError: /Users/aravisha/babel-core-issue1/packages/test-plugin/index.ts: Unexpected token, expected "{" (11:7)

   9 |   }
  10 |
> 11 |   run(): void {
     |        ^
  12 |     console.log("Running test plugin");
  13 |   }
  14 | }
packages/test-plugin/index.ts (11:7)
    at constructor (/Users/aravisha/babel-core-issue1/node_modules/@babel/parser/lib/index.js:345:19)
    at Parser.raise (/Users/aravisha/babel-core-issue1/node_modules/@babel/parser/lib/index.js:3199:19)
```

### Current and expected behavior

The build should work fine without breaking.

### Additional context

Note:  Without any other changes if we downgrade the @babel/core changes to 1 patch below (7.22.10) the build works fine.

I also get the below error sometimes especially the first time and the next time I get `Unexpected token, expected "{"` 

```
[!] (plugin babel) SyntaxError: /Users/aravisha/babel-core-issue1/packages/core/index.ts: Support for the experimental syntax 'flow' isn't currently enabled (1:8):
> 1 | export type { BaseConfig } from './config.js';
    |        ^
Add @babel/preset-flow (https://github.com/babel/babel/tree/main/packages/babel-preset-flow) to the 'presets' section of your Babel config to enable transformation.
If you want to leave it as-is, add @babel/plugin-syntax-flow (https://github.com/babel/babel/tree/main/packages/babel-plugin-syntax-flow) to the 'plugins' section to enable parsing.
packages/core/index.ts (1:7)`
```

Also there is also some caching issues. Sometimes I need to delete the node_modules and should do `yarn cache clean` to revert to my original state. Just changing the code is not reflecting the change.

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #15916
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

The re-entrancy check was breaking ESM config loading, because given that ESM loading is asynchronous it was possible for two `imports()` to be running at the same time.

Given that re-entrancy problems causing infinite config loading loops can only happen with require hooks, this PR moves the check to only apply to CJS config loading.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15923"></a>

### CODE REVIEW

File: packages/babel-core/test/config-loading.js [hunk 1]
```diff
@@ -166,6 +168,36 @@ describe("@babel/core config loading", () => {
     });
   });
 
+  describe("loadPartialConfigAsync", () => {
+    // https://github.com/babel/babel/issues/15916
+    (supportsESM ? it : it.skip)(
```
Comment (JLHwung) [inline packages/babel-core/test/config-loading.js:173]: Nit: may also use `itESM` from `$repo-utils`.

**Review by JLHwung — APPROVED** [review]

File: packages/babel-core/test/config-loading.js (hunk 1 above)
Comment (nicolo-ribaudo) [reply packages/babel-core/test/config-loading.js:173]: itESM is for when we are compiling to ESM, while in this case I want the test to run also when compiling to CJS if ESM config files are supported.

Comment (JLHwung) [reply packages/babel-core/test/config-loading.js:173]: You are right, I didn't figure out it was actually `itGte("12")`.

**Review by liuxingbaoyu — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
Releasing

**Pull request merged**